### PR TITLE
PIM-4690: Refacoring of the category reader

### PR DESCRIPTION
**Pull request by solivier**
| Q                  | A
| ------------------ | ---
| Migration script?  | n
| Behats             | running
| Specs & Static     | y
| Changelog updated  | y

### CODE REVIEW

File: src/Pim/Bundle/BaseConnectorBundle/Reader/ORM/CategoryReader.php [hunk 1]
```diff
@@ -30,18 +42,43 @@ public function __construct(EntityRepository $categoryRepository)
...
+    /**
+     * {@inheritdoc}
+     */
+    public function setStepExecution(StepExecution $stepExecution)
+    {
+        $this->stepExecution = $stepExecution;
+    }
+
+    /**
+     * {@inheritdoc}
```
Comment (willy-ahva) [inline src/Pim/Bundle/BaseConnectorBundle/Reader/ORM/CategoryReader.php]: Seems to not be inherited

File: src/Pim/Bundle/BaseConnectorBundle/Reader/ORM/CategoryReader.php [hunk 2]
```diff
@@ -30,18 +42,43 @@ public function __construct(EntityRepository $categoryRepository)
...
+     */
+    protected function getResults()
+    {
+        return new \ArrayIterator($this->categoryRepository->getOrderedAndSortedByTreeCategories());
```
Comment (jjanvier) [inline src/Pim/Bundle/BaseConnectorBundle/Reader/ORM/CategoryReader.php]: weird to have an array iterator on something that is already a collection...

File: src/Pim/Bundle/BaseConnectorBundle/Reader/ORM/CategoryReader.php [hunk 3]
```diff
@@ -12,17 +15,26 @@
...
-     * @var EntityRepository
-     */
+    /** @var CategoryRepositoryInterface */
     protected $categoryRepository;
 
+    /** @var StepExecution */
+    protected $stepExecution;
+
+    /** @var bool Checks if all categories are sent to the processor */
+    protected $executed = false;
```
Comment (willy-ahva) [inline src/Pim/Bundle/BaseConnectorBundle/Reader/ORM/CategoryReader.php]: $isExecuted

**Comment by jjanvier** [comment]
no ticket in commits, no ticket in the PR
don't understand why you need that

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/CategoryRepository.php [hunk 4]
```diff
@@ -363,4 +363,15 @@ public function search($treeRootId, $criterias)
 
         return $queryBuilder->getQuery()->getResult();
     }
+
+    /**
+     * @return ArrayCollection
+     */
+    public function getOrderedAndSortedByTreeCategories()
```
Comment (sumbobyboys) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/CategoryRepository.php]: inheritdoc

**Pull request closed without merging**